lib/luacrypto_shash.c

```diff
@@ -0,0 +1,292 @@
+* Low-level Lua interface to the Linux Kernel Crypto API for synchronous
+* message digest (hash) algorithms, including HMAC.
+*
+* This module provides a `new` function to create SHASH transform objects,
+* which can then be used for various hashing operations.
+*
+* @module crypto_shash
+*/
+
+#define pr_fmt(fmt) KBUILD_MODNAME ": " fmt
+
+#include <linux/kernel.h>
+#include <linux/module.h>
+#include <linux/crypto.h>
+#include <crypto/hash.h>
+#include <linux/err.h>
+#include <linux/slab.h>
+
+#include <lua.h>
+#include <lualib.h>
+#include <lauxlib.h>
+#include <lunatik.h>
+
+LUNATIK_PRIVATECHECKER(luacrypto_shash_check_tfm, struct shash_desc *);
+
+static void luacrypto_shash_release(void *private)
+{
+	struct crypto_shash *tfm;
+	struct shash_desc *sdesc = (struct shash_desc *)private;
+
+	if (sdesc) {
+		tfm = sdesc->tfm;
+		lunatik_free(sdesc);
+		if (tfm)
+			crypto_free_shash(tfm);
+	}
+}
+
+/***
+* SHASH object methods.
+* These methods are available on SHASH objects created by `crypto_shash.new()`.
+* @see crypto_shash.new
+* @type crypto_shash
+*/
+
+/***
+* Gets the digest size (output length) of the hash algorithm.
+* @function crypto_shash:digestsize
+* @treturn integer The digest size in bytes.
+*/
+static int luacrypto_shash_digestsize(lua_State *L) {
+	struct shash_desc *sdesc = luacrypto_shash_check_tfm(L, 1);
+	lua_pushinteger(L, crypto_shash_digestsize(sdesc->tfm));
+	return 1;
+}
+
+/***
+* Sets the key for the SHASH transform (used for HMAC).
+* @function crypto_shash:setkey
+* @tparam string key The key to use for HMAC.
+* @raise Error if setting the key fails.
+*/
+static int luacrypto_shash_setkey(lua_State *L) {
+	struct shash_desc *sdesc = luacrypto_shash_check_tfm(L, 1);
+	size_t keylen;
+	const char *key = luaL_checklstring(L, 2, &keylen);
+	lunatik_try(L, crypto_shash_setkey, sdesc->tfm, key, keylen);
+	return 0;
+}
+
+/***
+* Computes the hash of the given data in a single operation.
+* For HMAC, `setkey()` must have been called first.
+* This function initializes, updates, and finalizes the hash calculation.
+* @function crypto_shash:digest
+* @tparam string data The data to hash.
+* @treturn string The computed digest (hash output).
+* @raise Error on failure (e.g., allocation error, crypto API error).
+*/
+static int luacrypto_shash_digest(lua_State *L) {
+	struct shash_desc *sdesc = luacrypto_shash_check_tfm(L, 1);
+	size_t datalen;
+	const char *data = luaL_checklstring(L, 2, &datalen);
+	unsigned int digestsize = crypto_shash_digestsize(sdesc->tfm);
+	u8 *digest_buf = lunatik_checkalloc(L, digestsize);
+
+	lunatik_try(L, crypto_shash_digest, sdesc, data, datalen, digest_buf);
+	lua_pushlstring(L, (const char *)digest_buf, digestsize);
+	lunatik_free(digest_buf);
+	return 1;
+}
+
+/***
+* Initializes a multi-part hash operation.
+* This must be called before using `update()` or `final()`.
+* @function crypto_shash:init
+* @raise Error on failure.
+*/
+static int luacrypto_shash_init_method(lua_State *L) {
```

Comment:
Is there a reason to not doing it automatically on instantiation? I don't see the use case for calling new() and not calling init()

Reply:
`init` is required for multi-part operation; but when using single operations like `digest`, initialization is handled internally. I tried adding `lunatik_try(L, crypto_shash_init, sdesc);` within `new`, and running hkdf test caused system hang without any log in dmesg.

Reply:
Ack